server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefIterator;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.compress.CompressorFactory;
+import org.elasticsearch.common.recycler.Recycler;
+import org.elasticsearch.common.util.PageCacheRecycler;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.zip.DataFormatException;
+import java.util.zip.Inflater;
+
+public class TransportDecompressor implements Closeable {
+
+    private final Inflater inflater;
+    private final PageCacheRecycler recycler;
+    private final ArrayDeque<Recycler.V<byte[]>> pages;
+    private int pageOffset = PageCacheRecycler.BYTE_PAGE_SIZE;
+    private boolean hasReadHeader = false;
+
+    public TransportDecompressor(PageCacheRecycler recycler) {
+        this.recycler = recycler;
+        inflater = new Inflater(true);
+        pages = new ArrayDeque<>(4);
+    }
+
+    public int decompress(BytesReference bytesReference) throws IOException {
+        int bytesConsumed = 0;
+        if (hasReadHeader == false) {
+            if (CompressorFactory.COMPRESSOR.isCompressed(bytesReference) == false) {
+                int maxToRead = Math.min(bytesReference.length(), 10);
+                StringBuilder sb = new StringBuilder("stream marked as compressed, but no compressor found, first [")
+                    .append(maxToRead).append("] content bytes out of [").append(bytesReference.length())
+                    .append("] readable bytes with message size [").append(bytesReference.length()).append("] ").append("] are [");
+                for (int i = 0; i < maxToRead; i++) {
+                    sb.append(bytesReference.get(i)).append(",");
+                }
+                sb.append("]");
+                throw new IllegalStateException(sb.toString());
+            }
+            hasReadHeader = true;
+            int headerLength = CompressorFactory.COMPRESSOR.headerLength();
+            bytesReference = bytesReference.slice(headerLength, bytesReference.length() - headerLength);
+            bytesConsumed += headerLength;
+        }
+
+        BytesRefIterator refIterator = bytesReference.iterator();
+        BytesRef ref;
+        while ((ref = refIterator.next()) != null) {
+            inflater.setInput(ref.bytes, ref.offset, ref.length);
+            bytesConsumed += ref.length;
+            boolean continueInflating = true;
+            while (continueInflating) {
+                if (pageOffset == PageCacheRecycler.BYTE_PAGE_SIZE) {
+                    pageOffset = 0;
+                    pages.add(recycler.bytePage(false));
```

Comment:
It seems we're wasting a page here every time we deflate `0` bytes (as it still gets allocated). Not sure if that matters much ever but:
Maybe simplify the logic to something like:

```diff
diff --git a/server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java b/server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java
index 69799c33d3d..c13bbf78008 100644
--- a/server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java
+++ b/server/src/main/java/org/elasticsearch/transport/TransportDecompressor.java
@@ -75,13 +75,25 @@ public class TransportDecompressor implements Closeable {
             bytesConsumed += ref.length;
             boolean continueInflating = true;
             while (continueInflating) {
-                if (pageOffset == PageCacheRecycler.BYTE_PAGE_SIZE) {
+                final Recycler.V<byte[]> page;
+                final boolean isNewPage = pageOffset == PageCacheRecycler.BYTE_PAGE_SIZE;
+                if (isNewPage) {
                     pageOffset = 0;
-                    pages.add(recycler.bytePage(false));
+                    page = recycler.bytePage(false);
+                } else {
+                    page = pages.getLast();
                 }
-                byte[] output = pages.getLast().v();
+                byte[] output = page.v();
                 try {
                     int bytesInflated = inflater.inflate(output, pageOffset, PageCacheRecycler.BYTE_PAGE_SIZE - pageOffset);
+                    if (isNewPage) {
+                        if (bytesInflated == 0) {
+                            page.close();
+                            pageOffset = PageCacheRecycler.BYTE_PAGE_SIZE;
+                        } else {
+                            pages.add(page);
+                        }
+                    }
                     pageOffset += bytesInflated;
                 } catch (DataFormatException e) {
                     throw new IOException("Exception while inflating bytes", e);
```

?